### DISCUSSION TITLE
为什么VUE版本的穿梭框（Transfer）没有提供异步数据加载的DEMO？而React版本有

### DISCUSSION THREAD

**Discussion by lyqyq2001** [ant-design/ant-design#51641] [General]
VUE版本没有办法做到支持异步数据加载吗？得在VUE里写JSX才行吗？
这是React版本的DEMO链接https://codesandbox.io/s/objective-wing-6iqbx

**Comment by petercat-assistant** [comment]
在 Ant Design Vue 中，虽然官方文档可能没有直接提供异步数据加载的示例，但实际上是可以实现的。Vue 版本的穿梭框（Transfer）组件同样支持异步数据加载。你不需要在 Vue 中写 JSX 来实现这一点。

你可以通过以下步骤来实现异步数据加载：

1. **使用 Vue 的生命周期钩子**：在组件的 `created` 或 `mounted` 钩子中发起异步请求以获取数据。

2. **使用 Vue 的响应式数据**：将获取到的数据存储在组件的 `data` 中，并将其绑定到穿梭框组件的 `dataSource` 属性。

3. **处理异步请求**：可以使用 `axios` 或 `fetch` 等工具来发起异步请求。

下面是一个简单的示例代码：

```vue
<template>
  <a-transfer
    :data-source="mockData"
    :target-keys="targetKeys"
    @change="handleChange"
    :render="item => item.title"
  />
</template>

<script>
import { Transfer } from 'ant-design-vue';
import axios from 'axios';

export default {
  components: {
    ATransfer: Transfer,
  },
  data() {
    return {
      mockData: [],
      targetKeys: [],
    };
  },
  created() {
    this.fetchData();
  },
  methods: {
    fetchData() {
      axios.get('your-api-endpoint')
        .then(response => {
          this.mockData = response.data.map(item => ({
            key: item.id,
            title: item.name,
            description: item.description,
          }));
        })
        .catch(error => {
          console.error('Failed to fetch data:', error);
        });
    },
    handleChange(nextTargetKeys) {
      this.targetKeys = nextTargetKeys;
    },
  },
};
</script>
```

在这个示例中，`fetchData` 方法用于从 API 获取数据，并将其格式化为穿梭框组件所需的格式。

如果你有任何其他问题或需要进一步的帮助，请在评论中描述你的问题并 @petercat-assistant 来开始与我对话。

**Comment by afc163** [comment]
问错地方了。

**Comment by lyqyq2001** [comment]
@petercat-assistant 
以下是问题的描述：在使用分页的穿梭框后，我绑定的data-source是pagesize为1的数据；假设有10条数据，如果我将其中的2条数据移到右边去，然后我再点击下一页去请求pagesize为2的数据，那么此时，我之前选择的2条数据在右边就消失了，因为绑定的data-source不一样了

**Comment by petercat-assistant** [comment]
在使用分页的穿梭框（Transfer）组件时，遇到的问题是由于分页导致的 `data-source` 变化，使得已选择的数据在切换页面后消失。要解决这个问题，你可以考虑以下方法：

1. **维护一个全局的选中状态**：在组件的 `data` 中维护一个全局的选中状态数组，用于存储所有选中的数据项的 `key`。每次分页数据变化时，更新 `data-source` 的同时，不要清空这个全局选中状态。

2. **在 `change` 事件中更新全局选中状态**：在 `Transfer` 组件的 `change` 事件中，更新全局选中状态数组。确保每次选中或取消选中时，都会正确地更新这个数组。

3. **在 `render` 过程中合并选中状态**：在渲染 `Transfer` 组件时，将全局选中状态与当前页的 `data-source` 合并，以确保选中的项始终显示在右侧。

以下是一个示例代码：

```vue
<template>
  <a-transfer
    :data-source="currentDataSource"
    :target-keys="targetKeys"
    @change="handleChange"
    :render="item => item.title"
  />
  <button @click="nextPage">Next Page</button>
</template>

<script>
import { Transfer } from 'ant-design-vue';

export default {
  components: {
    ATransfer: Transfer,
  },
  data() {
    return {
      allData: [], // 全部数据
      currentDataSource: [], // 当前页的数据
      targetKeys: [], // 选中的数据key
      currentPage: 1,
      pageSize: 1,
    };
  },
  created() {
    this.fetchData();
  },
  methods: {
    fetchData() {
      // 模拟获取数据
      this.allData = Array.from({ length: 10 }, (_, i) => ({
        key: i.toString(),
        title: `Item ${i + 1}`,
      }));
      this.updateCurrentDataSource();
    },
    updateCurrentDataSource() {
      const start = (this.currentPage - 1) * this.pageSize;
      const end = start + this.pageSize;
      this.currentDataSource = this.allData.slice(start, end);
    },
    handleChange(nextTargetKeys) {
      this.targetKeys = nextTargetKeys;
    },
    nextPage() {
      this.currentPage += 1;
      this.updateCurrentDataSource();
    },
  },
};
</script>
```

在这个示例中，`allData` 存储了所有的数据，而 `currentDataSource` 仅存储当前页的数据。`targetKeys` 用于存储所有选中的数据项的 `key`。通过 `nextPage` 方法来切换分页，并更新 `currentDataSource`。

如果你有任何其他问题或需要进一步的帮助，请在评论中描述你的问题并 @petercat-assistant 来开始与我对话。
